lib/r2d2/google_pay_token.rb

```diff
@@ -33,22 +41,104 @@ def decrypt(private_key_pem)
     private
 
     def verify_and_parse_message
-      digest = OpenSSL::Digest::SHA256.new
+      case protocol_version
+      when 'ECv1'
+        verify_and_parse_message_ecv1
+      when 'ECv2'
+        verify_and_parse_message_ecv2
+      else
+        raise ArgumentError, "unknown protocolVersion #{protocol_version}"
+      end
+    end
+
+    def verify_and_parse_message_ecv1
       signed_bytes = to_length_value(
         'Google',
         recipient_id,
         protocol_version,
         signed_message
       )
-      verified = verification_keys['keys'].any? do |key|
-        next if key['protocolVersion'] != protocol_version
 
-        ec = OpenSSL::PKey::EC.new(Base64.strict_decode64(key['keyValue']))
-        ec.verify(digest, Base64.strict_decode64(signature), signed_bytes)
-      end
+      verified = valid_key_signatures?(
+        verification_keys,
+        [signature],
+        signed_bytes
+      )
 
       raise SignatureInvalidError unless verified
       JSON.parse(signed_message)
     end
+
+    def verify_and_parse_message_ecv2
+      raise SignatureInvalidError, 'intermediate certificate is expired' if intermediate_key_expired?
+      raise SignatureInvalidError, 'no valid signature of intermediate key' unless intermediate_key_signature_verified?
+      raise SignatureInvalidError, 'signature of signedMessage does not match' unless payload_signature_verified?
+
+      JSON.parse(signed_message)
+    end
+
+    ### ECv2 Methods ###
+    def intermediate_key_signature_verified?
+      intermediate_signatures = intermediate_signing_key['signatures']
+      signed_bytes = [sender_id, protocol_version, intermediate_signing_key['signedKey']].map do |str|
+        [str.length].pack('V') + str
+      end.join
+
+      # Check that at least one of the root keys signed the intermediate
+      valid_key_signatures?(
+        verification_keys,
+        intermediate_signatures,
+        signed_bytes
+      )
+    end
+
+    def payload_signature_verified?
+      signed_string_message = [sender_id, ecv2_recipient_id, protocol_version, signed_message].map do |str|
+        [str.length].pack('V') + str
+      end.join
+
+      # Check that the intermediate key signed the message
+      pkey = OpenSSL::PKey::EC.new(Base64.strict_decode64(intermediate_signing_key_signed_key['keyValue']))
+      valid_key_signatures?(
+        [pkey],
+        [signature],
+        signed_string_message
+      )
+    end
+
+    def valid_key_signatures?(signing_keys, signatures, signed)
+      signing_keys.product(signatures).any? do |key, sig|
+        key.verify(OpenSSL::Digest::SHA256.new, Base64.strict_decode64(sig), signed)
```

Comment:
Don't use [deprecated constants](https://github.com/rubocop-hq/rubocop/pull/7950):
```suggestion
        key.verify(OpenSSL::Digest.new("SHA256"), Base64.strict_decode64(sig), signed)
```